src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
                             _rcwCache.Remove(identity);
                         }
                     }
+                }
+
+                if (wrapperMaybe is not null)
+                {
+                    retValue = wrapperMaybe;
 
-                    if (wrapperMaybe is not null)
+                    using (_lock.EnterScope())
                     {
-                        retValue = wrapperMaybe;
-                        NativeObjectWrapper wrapper = NativeObjectWrapper.Create(
-                            identity,
-                            inner,
-                            this,
-                            retValue,
-                            flags);
-                        if (!s_rcwTable.TryAdd(retValue, wrapper))
+                        if (!s_rcwTable.TryGetValue(retValue, out NativeObjectWrapper? wrapper))
                         {
-                            wrapper.Release();
-                            throw new NotSupportedException();
+                            // Exit the lock before we create the native object wrapper
+                            // as we may execute arbitrary user code.
+                            _lock.Exit();
+                            try
+                            {
+                                NativeObjectWrapper newWrapper = NativeObjectWrapper.Create(
+                                    identity,
+                                    inner,
+                                    this,
+                                    retValue,
+                                    flags);
+
+                                wrapper = s_rcwTable.GetValue(retValue, _ => newWrapper);
```

Comment:
Can this use `TryAdd` like before to avoid the delegate allocation?

Reply:
`TryAdd` doesn't get us the current value and `TryAdd` + `TryGetValue` would be a race condition (if a GC happened in the middle, we could need to do `TryAdd` again).

I'd like to use `GetValue` that takes a creation delegate, but I'm trying to avoid leaking the unused `NativeObjectWrapper` and relying on the finalizer to clean it up.

Reply:
I think you can do:
```
if (s_rcwTable.TryAdd(retValue, newWrapper))
{
   wrapper = newWrapper;
}
else
{
   wrapper = s_rcwTable.GetValue(retValue, _ => newWrapper);
}
```
It avoids the delegate allocation for the common case.

Reply:
> "It avoids the delegate allocation for the common case."

@jkotas I remember this also popped up before in an unrelated PR, I don't think doing that is enough to ensure we skip the allocation, unfortunately, See [here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAlAVwDsAXASwFsZCBhCWgB2oBsYoBlbgN2pgYAZwDcAWABQ0jAGY8AJhyNpAb2k4tOVlGr8AhpRh58SHLAMATCOQ4BPFbavUatgxwDqMAwGsAKgbAXAA8EMAAVjBglAA0OOFRMQB8OCIA+lBgAO6BwSYAvDjkMNkAFACUklLaOJra8gmR0ZQ4ALJliS0WMJQAah6kMPFdMcWlnlAGrKzcFfVaGjW12qOt2VMz3NUr2tQAZjhlGVm5QVyE/lD2AIJWVmWw/YPD49mT07NQFfPLu0u7FYbT7cHBFErvTZfHaAgC+CxWMA4IhgCNqAMB2mBWygYLSmRyeQuAHFegMOENHmSXvF0mDUhCPjiqmjtPC/isMAB2HDY6EI9mwoA==), Roslyn has the bad habit of eagerly allocating the display class and using that for locals even if you don't actually need to use the delegate closing over it at runtime. To skip the allocation we need to move the `GetValue` path into a separate small stub method just doing that, to ensure the main block cannot be lowered to use that same display class.

That is, something like this:

```csharp
if (s_rcwTable.TryAdd(retValue, newWrapper))
{
    wrapper = newWrapper;
}
else
{
    static object RcwTableGetValue(object retValue, object newWrapper)
        => s_rcwTable.GetValue(retValue, _ => newWrapper);

    wrapper = RcwTableGetValue(retValue, newWrapper);
}
```